Add 'symbol_source' config

ttreyer opened the pull request:
Let the user decide whether to use DebugInfo to determine `uprobe` location, or use the Symbol Table.

The new `symbol_source` config takes the following two values:
- `dwarf`: uses the DebugInfo, which yield more accurate stack traces. However, if the DebugInfo was rewritten by a post-linkage optimisation tool (like BOLT or AutoFDO), we might get wrong addresses for the probe location.
- `symbol_table`: uses the target's Symbol Table, which won't suffer from improperly rewritten DebugInfo.

This will close #3493.

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests
  - [x] Test setting the config
  - [x] Test the config impact

jordalgo commented on CHANGELOG.md:
```diff
@@ -29,6 +29,8 @@ and this project adheres to
   - [#3390](https://github.com/bpftrace/bpftrace/pull/3390/)
 - Add `has_key` function for maps
   - [#3358](https://github.com/bpftrace/bpftrace/pull/3358)
+- Add `symbol_source` config to source uprobe locations from either DWARF or the Symbol Table
```
Naming question: is `uprobe_symbol_source` to long?

jordalgo commented on man/adoc/bpftrace.adoc:
```diff
@@ -3563,6 +3563,16 @@ Default: `..`
+
+Default: `dwarf`
+
+Choose how bpftrace will resolve `uprobe`s' location.
```
I would also call out that we default to `symbol_table` if LLDB is not available.

jordalgo commented on man/adoc/bpftrace.adoc:
```diff
@@ -3563,6 +3563,16 @@ Default: `..`
+Choose how bpftrace will resolve `uprobe`s' location.
+
+Available options:
+- `dwarf` - locate probe using DebugInfo, it will give better stack trace output. bpftrace will fall back to the Symbol Table if it can't locate the probe using DebugInfo.
```
> fall back to the Symbol Table if it can't locate the probe using DebugInfo.

Where does this happen?

ttreyer replied:
Right [here](https://github.com/bpftrace/bpftrace/pull/3504/files#diff-f6bea5bb780ac5ca59678079c8010b81f5b8eb83cbe90f77e134fd6f2a742a32R168-R170).

jordalgo reviewed (changes requested):
Code looks pretty good. Maybe good to also add a runtime test (if that's possible).

ttreyer commented on CHANGELOG.md:
```diff
@@ -29,6 +29,8 @@ and this project adheres to
   - [#3390](https://github.com/bpftrace/bpftrace/pull/3390/)
 - Add `has_key` function for maps
   - [#3358](https://github.com/bpftrace/bpftrace/pull/3358)
+- Add `symbol_source` config to source uprobe locations from either DWARF or the Symbol Table
```
It would be fine, but this settings is also going to apply to kprobes in #3301. Do you have a suggestion fitting for both?

jordalgo replied:
Oh, ok. Didn't know kprobes were going to be using it too. This is fine then.

ajor replied:
I'm not so bothered about the name but just want to clarify what the plans are for these config options.

Inlined kprobes are already going to be controlled by the probe_inline config option, right? For non-inlined kprobes, I'm not sure there's any benefit to using LLDB's locations. The reason for uprobes was that the kernel doesn't provide correct stack traces otherwise, but that's not the case for kprobes.

viktormalik replied:
To second what @ajor said, would it make sense to extend the `probe_inline` option to choose whether to use DWARF or symbol table for placing the probes? And use symbol table for non-inlined uprobes?

ajor replied:
There is still a reason to have two separate config options - we want to control these two things independently:
- Uprobes/kprobes tracing inlined functions
- Uprobes using debug info to pick attach points for non-inlined functions (to fix ustacks: #646)

The second one is needed because we've come across a case where BOLT (or similar) messes up the DWARF info for non-inlined functions, so we need a way to fall back to the symbol table in that case.

We shouldn't be using DWARF info for non-inlined kprobes anyway, so this config option only needs to apply to uprobes.

viktormalik replied:
Ah, ok, thanks for the clarification.

jordalgo commented:
Looks good! One last thing (and maybe this needs to be a separate PR) but can we detect if users are setting an offset themselves or using a glob for their uprobe and, in the latter case, suggest using this new config if `check_alignment` fails (https://github.com/bpftrace/bpftrace/blob/2fa7894baf76934b761772962a8154485c6ba44c/src/attached_probe.cpp#L405).

Also was a runtime test too much of a PITA?

ttreyer commented:
> Looks good! One last thing (and maybe this needs to be a separate PR) but can we detect if users are setting an offset themselves or using a glob for their uprobe and, in the latter case, suggest using this new config if `check_alignment` fails (
> 
> https://github.com/bpftrace/bpftrace/blob/2fa7894baf76934b761772962a8154485c6ba44c/src/attached_probe.cpp#L405
> 
> ).

Good suggestion! I've added it as a hint.

> Also was a runtime test too much of a PITA?

My bad, I've added two to test both with and without `probe_inline` enabled.

jordalgo commented on man/adoc/bpftrace.adoc:
```diff
@@ -3563,6 +3563,19 @@ Default: `..`
+Choose how bpftrace will resolve all `uprobe` symbol locations.
+
+Available options:
+- `dwarf` - locate uprobes using DebugInfo, which yields more accurate stack traces (`ustack`). Fall back to the Symbol Table if it can't locate the probe using DebugInfo.
```
nit: Insert a line break above this list, it's not getting formatted correctly in the markdown.

jordalgo commented on src/attached_probe.cpp:
```diff
@@ -401,13 +405,17 @@ static void check_alignment(std::string &path,
+        auto msg = "Could not add " + probetypeName(type) +
+                   " into middle of instruction: " + tmp +
+                   std::string{ hint_unsafe };
+        if (orig_name.find('*') != std::string::npos)
```
Should we also add this hint to the other failing case statements below?

jordalgo replied:
@ttreyer I meant using `hint_symbol_source` for the AlignState::Fail and AlignState::NotSupp - or do those not apply?

ttreyer replied:
I understand `AlignState::Fail` and `AlignState::NotSupp` to be errors returned by the disassembler, not necessary an issue with the address itself. That's why I'm not convinced that the hint applies to both of them.

jordalgo replied:
Would these failure occur if the user switched to `symbol_table` mode?

ttreyer replied:
Yes, the failure would still occur.

jordalgo replied:
Ok cool. I'm satisfied then 😄

jordalgo commented on tests/runtime/dwarf:
```diff
@@ -74,6 +74,22 @@ EXPECT foo1->d = 9abcdef0
+     uprobe:./testprogs/inline_function:square { @count++ }
+     uretprobe:./testprogs/inline_function:main { exit() }
+EXPECT @count: 1
+REQUIRES_FEATURE dwarf
```
why does this one require dwarf?

ttreyer replied:
It requires `liblldb` to test that the `symbol_source` config takes precedence over `probe_inline`.

jordalgo commented:
Looks like maybe one of the new added tests needs to fixed for AOT or added to the skip list.

ttreyer commented:
I've added it to the AOT skip list, as Dwarf is currently not compatible with AOT.

fbs commented:
Somewhat uninformed question:

Does it make sense for this to be a 'global' toggle, or would you want to be able to do this per probe? Or maybe a global and per probe?

ttreyer commented:
> Does it make sense for this to be a 'global' toggle, or would you want to be able to do this per probe? Or maybe a global and per probe?

It would make sense to have config per-probe instead of always being global. But `bpftrace` doesn't support per-probe config currently.
